File: src/audio.rs
use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
use std::sync::{Arc, Mutex};

use crate::pulse;

const BLOCK_FRAMES: usize = 512;

struct Shared {
    head: AtomicUsize,
    tail: AtomicUsize,
    terminate: AtomicBool,
    error: Mutex<String>,
    ring: Vec<AtomicUsize>,
    blocks: Vec<Mutex<Vec<f64>>>,
    mask: usize,
}

pub struct Audio {
    shared: Arc<Shared>,
    left: Vec<f64>,
    right: Vec<f64>,
    channels: usize,
    thread: Option<std::thread::JoinHandle<()>>,
}

fn next_pow2(v: usize) -> usize {
    let mut p = 1usize;
    while p < v {
        p <<= 1;
    }
    p
}

impl Audio {
    pub fn new(max_frames: usize) -> Self {
        let max_frames = max_frames.max(BLOCK_FRAMES);
        let nblocks = next_pow2((max_frames / BLOCK_FRAMES).max(2) + 2);
        let shared = Shared {
            head: AtomicUsize::new(0),
            tail: AtomicUsize::new(0),
            terminate: AtomicBool::new(false),
            error: Mutex::new(String::new()),
            ring: (0..nblocks)
                .map(|_| AtomicUsize::new(usize::MAX))
                .collect(),
            blocks: (0..nblocks)
                .map(|_| Mutex::new(vec![0.0; BLOCK_FRAMES * 2]))
                .collect(),
            mask: nblocks - 1,
        };
        Audio {
            shared: Arc::new(shared),
            left: vec![0.0; max_frames],
            right: vec![0.0; max_frames],
            channels: 2,
            thread: None,
        }
    }

    pub fn start(&mut self, source: &str, rate: u32, channels: u32) {
        let channels = channels.clamp(1, 2);
        self.channels = channels as usize;
        let shared = self.shared.clone();
        let source = source.to_string();
        self.thread = Some(std::thread::spawn(move || {
            capture(shared, source, rate, channels);
        }));
    }

    pub fn consume(&mut self) -> (usize, Option<&[f64]>, Option<&[f64]>) {
        let sh = &self.shared;
        let cap = self.left.len();
        let mut frames = 0usize;

        loop {
            let head = sh.head.load(Ordering::Acquire);
            let tail = sh.tail.load(Ordering::Relaxed);
            if head == tail || frames >= cap {
                break;
            }
            let slot = tail & sh.mask;
            let n = sh.ring[slot].load(Ordering::Relaxed);
            if n == usize::MAX || n > BLOCK_FRAMES {
                break;
            }
            // Blocks are variable-length now (partial flushes); leave the
            // block queued if it would overflow this frame's buffer.
            if n > 0 && frames + n > cap {
                break;
            }
            let b = sh.blocks[slot].lock().unwrap();
            let base = frames;
            let mut j = 0;
            for i in 0..n {
                self.left[base + i] = b[j];
                self.right[base + i] = b[j + 1];
                j += 2;
            }
            drop(b);
            let _ = sh.ring[slot].store(usize::MAX, Ordering::Relaxed);
            sh.tail.fetch_add(1, Ordering::Release);
            frames += n;
        }

        let n = frames;
        let left = if n > 0 { Some(&self.left[..n]) } else { None };
        let right = if self.channels > 1 && n > 0 {
            Some(&self.right[..n])
        } else {
            None
        };
        (n, left, right)
    }

    pub fn failed(&self) -> bool {
        self.shared.terminate.load(Ordering::SeqCst)
    }

    pub fn error(&self) -> String {
        self.shared.error.lock().unwrap().clone()
    }

    pub fn stop(&mut self) {
        self.shared.terminate.store(true, Ordering::SeqCst);
        if let Some(t) = self.thread.take() {
            let _ = t.join();
        }
    }
}

fn dev_name(p: &mut pulse::Pulse, source: &str) -> Option<String> {
    if source.is_empty() || source == "auto" || source == "default" {
        p.default_monitor().ok()
    } else {
        Some(source.to_string())
    }
}

fn push_block(shared: &Arc<Shared>, staged: &[f64], staged_cnt: usize) {
    if staged_cnt == 0 || staged_cnt > BLOCK_FRAMES {
        return;
    }
    let head = shared.head.load(Ordering::Relaxed);
    let tail = shared.tail.load(Ordering::Acquire);
    if head - tail < shared.blocks.len() {
        let slot = head & shared.mask;
        let mut b = shared.blocks[slot].lock().unwrap();
        b[..staged_cnt * 2].copy_from_slice(&staged[..staged_cnt * 2]);
        let _ = shared.ring[slot].store(staged_cnt, Ordering::Relaxed);
        shared.head.fetch_add(1, Ordering::Release);
    }
    // Ring full: drop the newest samples (same policy as before).
}

fn capture(shared: Arc<Shared>, source: String, rate: u32, channels: u32) {
    let nch = channels.clamp(1, 2);
    let mut p = match pulse::Pulse::connect() {
        Ok(p) => p,
        Err(e) => {
            *shared.error.lock().unwrap() = e;
            shared.terminate.store(true, Ordering::SeqCst);
            return;
        }
    };

    let dev = match dev_name(&mut p, &source) {
        Some(d) => d,
        None => {
            *shared.error.lock().unwrap() =
                "pulse: could not resolve default sink monitor".to_string();
            shared.terminate.store(true, Ordering::SeqCst);
            return;
        }
    };

    let bytes_per_sec = nch as u64 * 2 * rate as u64;
    let fragsize = (5000 * bytes_per_sec / 1_000_000) as u32;

    let rec = match p.record(&dev, rate, nch as u8, fragsize) {
        Ok(r) => r,
        Err(e) => {
            *shared.error.lock().unwrap() = format!("pulse: record stream failed: {e}");
            shared.terminate.store(true, Ordering::SeqCst);
            return;
        }
    };
    let mut rec = rec;

    let nbytes_per_frame = nch as usize * 2;
    let mut raw = vec![0u8; BLOCK_FRAMES * nbytes_per_frame];
    let mut staged = vec![0.0f64; BLOCK_FRAMES * 2];
    let mut staged_cnt = 0usize;

    loop {
        match rec.read(&mut raw, &shared.terminate) {
            Ok(0) => {
                // Flush any staged tail before exiting so no audio is lost.
                if staged_cnt > 0 {
                    push_block(&shared, &staged, staged_cnt);
                }
                break;
            }
            Ok(n) => {
                let nframes = n / nbytes_per_frame;
                let mut f = 0;
                while f < nframes {
                    // Fill the staging block; push it as soon as it is
                    // full and keep going so loud/high-rate reads that
                    // exceed BLOCK_FRAMES are split across blocks instead
                    // of dropping the tail (old code `break`ed here).
                    while f < nframes && staged_cnt < BLOCK_FRAMES {
                        let l = i16::from_le_bytes([
                            raw[f * nbytes_per_frame],
                            raw[f * nbytes_per_frame + 1],
                        ]) as f64
                            / 32768.0;
                        let r = if nch >= 2 {
                            i16::from_le_bytes([
                                raw[f * nbytes_per_frame + 2],
                                raw[f * nbytes_per_frame + 3],
                            ]) as f64
                                / 32768.0
                        } else {
                            l
                        };
                        staged[staged_cnt * 2] = l;
                        staged[staged_cnt * 2 + 1] = r;
                        staged_cnt += 1;
                        f += 1;
                    }
                    if staged_cnt >= BLOCK_FRAMES {
                        push_block(&shared, &staged, staged_cnt);
                        staged_cnt = 0;
                    }
                }

                // Push partial blocks immediately instead of waiting for a
                // full 512 frames. A full block spans 512/rate seconds
                // (~64ms at 8kHz vs ~10ms at 48kHz); waiting for it made
                // low sample rates deliver audio in bursts every few
                // display frames, which collapsed the visual refresh rate
                // and made motion jumpy. Flushing each read keeps delivery
                // at the ~5ms Pulse fragment cadence for every rate.
                if staged_cnt > 0 {
                    push_block(&shared, &staged, staged_cnt);
                    staged_cnt = 0;
                }
            }
            Err(e) => {
                if !shared.terminate.load(Ordering::SeqCst) {
                    *shared.error.lock().unwrap() = e;
                    shared.terminate.store(true, Ordering::SeqCst);
                }
                break;
            }
        }
    }
}

File: src/audio.c
#include <pthread.h>
#include <stdatomic.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "audio.h"
#include "pulse.h"

#define BLOCK_FRAMES 512

typedef struct {
    _Atomic size_t head;
    _Atomic size_t tail;
    _Atomic int terminate;
    pthread_mutex_t err_mu;
    char error[1024];
    _Atomic size_t *ring;
    pthread_mutex_t *block_mu;
    double *blocks;
    size_t nblocks;
    size_t mask;
} Shared;

struct Audio {
    Shared *shared;
    double *left;
    double *right;
    size_t cap;
    size_t channels;
    pthread_t thread;
    int started;
    char want_source[256];
    unsigned want_rate;
    unsigned want_channels;
};

static size_t next_pow2(size_t v) {
    size_t p = 1;
    while (p < v)
        p <<= 1;
    return p;
}

Audio *audio_new(size_t max_frames) {
    if (max_frames < BLOCK_FRAMES)
        max_frames = BLOCK_FRAMES;
    size_t nb = max_frames / BLOCK_FRAMES;
    if (nb < 2)
        nb = 2;
    nb = next_pow2(nb + 2);
    Audio *a = calloc(1, sizeof *a);
    if (!a)
        return NULL;
    Shared *sh = calloc(1, sizeof *sh);
    if (!sh) {
        free(a);
        return NULL;
    }
    atomic_init(&sh->head, 0);
    atomic_init(&sh->tail, 0);
    atomic_init(&sh->terminate, 0);
    pthread_mutex_init(&sh->err_mu, NULL);
    sh->ring = calloc(nb, sizeof(size_t));
    sh->block_mu = calloc(nb, sizeof(pthread_mutex_t));
    sh->blocks = calloc(nb * BLOCK_FRAMES * 2, sizeof(double));
    if (!sh->ring || !sh->block_mu || !sh->blocks) {
        free(sh->ring);
        free(sh->block_mu);
        free(sh->blocks);
        free(sh);
        free(a);
        return NULL;
    }
    for (size_t i = 0; i < nb; i++) {
        atomic_init(&sh->ring[i], (size_t)-1);
        pthread_mutex_init(&sh->block_mu[i], NULL);
    }
    sh->nblocks = nb;
    sh->mask = nb - 1;
    a->shared = sh;
    a->cap = max_frames;
    a->left = calloc(max_frames, sizeof(double));
    a->right = calloc(max_frames, sizeof(double));
    a->channels = 2;
    if (!a->left || !a->right) {
        audio_free(a);
        return NULL;
    }
    return a;
}

void audio_free(Audio *a) {
    if (!a)
        return;
    audio_stop(a);
    Shared *sh = a->shared;
    if (sh) {
        for (size_t i = 0; i < sh->nblocks; i++)
            pthread_mutex_destroy(&sh->block_mu[i]);
        pthread_mutex_destroy(&sh->err_mu);
        free(sh->ring);
        free(sh->block_mu);
        free(sh->blocks);
        free(sh);
    }
    free(a->left);
    free(a->right);
    free(a);
}

static void set_error(Shared *sh, const char *msg) {
    pthread_mutex_lock(&sh->err_mu);
    snprintf(sh->error, sizeof sh->error, "%s", msg);
    pthread_mutex_unlock(&sh->err_mu);
}

static void push_block(Shared *sh, const double *staged, size_t cnt) {
    if (cnt == 0 || cnt > BLOCK_FRAMES)
        return;
    size_t head = atomic_load(&sh->head);
    size_t tail = atomic_load(&sh->tail);
    if (head - tail < sh->nblocks) {
        size_t slot = head & sh->mask;
        pthread_mutex_lock(&sh->block_mu[slot]);
        memcpy(sh->blocks + slot * BLOCK_FRAMES * 2, staged, cnt * 2 * sizeof(double));
        pthread_mutex_unlock(&sh->block_mu[slot]);
        atomic_store(&sh->ring[slot], cnt);
        atomic_fetch_add(&sh->head, 1);
    }
}

static void *capture_thread(void *arg) {
    Audio *a = arg;
    Shared *sh = a->shared;
    char source[256];
    snprintf(source, sizeof source, "%s", a->want_source);
    unsigned rate = a->want_rate;
    unsigned nch = a->want_channels;
    if (nch < 1)
        nch = 1;
    if (nch > 2)
        nch = 2;
    char err[512];
    Pulse *p = pulse_connect(err, sizeof err);
    if (!p) {
        set_error(sh, err);
        atomic_store(&sh->terminate, 1);
        return NULL;
    }
    char dev[512];
    if (!source[0] || !strcmp(source, "auto") || !strcmp(source, "default")) {
        if (!pulse_default_monitor(p, dev, sizeof dev, err, sizeof err)) {
            set_error(sh, "pulse: could not resolve default sink monitor");
            pulse_free(p);
            atomic_store(&sh->terminate, 1);
            return NULL;
        }
    } else {
        snprintf(dev, sizeof dev, "%s", source);
    }
    unsigned long long bytes_per_sec = (unsigned long long)nch * 2 * rate;
    unsigned fragsize = (unsigned)(5000 * bytes_per_sec / 1000000);
    Record *rec = pulse_record(p, dev, rate, nch, fragsize, err, sizeof err);
    if (!rec) {
        char msg[640];
        snprintf(msg, sizeof msg, "pulse: record stream failed: %s", err);
        set_error(sh, msg);
        atomic_store(&sh->terminate, 1);
        return NULL;
    }
    size_t bpf = nch * 2;
    uint8_t *raw = malloc(BLOCK_FRAMES * bpf);
    double *staged = malloc(BLOCK_FRAMES * 2 * sizeof(double));
    size_t staged_cnt = 0;
    if (!raw || !staged) {
        free(raw);
        free(staged);
        record_free(rec);
        atomic_store(&sh->terminate, 1);
        return NULL;
    }
    for (;;) {
        if (atomic_load(&sh->terminate))
            break;
        long n = record_read(rec, raw, BLOCK_FRAMES * bpf,
                             (const volatile int *)&sh->terminate, err, sizeof err);
        if (n <= 0) {
            if (n < 0 && !atomic_load(&sh->terminate))
                set_error(sh, err);
            if (staged_cnt > 0)
                push_block(sh, staged, staged_cnt);
            atomic_store(&sh->terminate, 1);
            break;
        }
        size_t nframes = (size_t)n / bpf;
        size_t f = 0;
        while (f < nframes) {
            while (f < nframes && staged_cnt < BLOCK_FRAMES) {
                int16_t l = (int16_t)((unsigned)raw[f * bpf] |
                                      ((unsigned)raw[f * bpf + 1] << 8));
                int16_t rr;
                if (nch >= 2) {
                    rr = (int16_t)((unsigned)raw[f * bpf + 2] |
                                   ((unsigned)raw[f * bpf + 3] << 8));
                } else {
                    rr = l;
                }
                staged[staged_cnt * 2] = (double)l / 32768.0;
                staged[staged_cnt * 2 + 1] = (double)rr / 32768.0;
                staged_cnt++;
                f++;
            }
            if (staged_cnt >= BLOCK_FRAMES) {
                push_block(sh, staged, staged_cnt);
                staged_cnt = 0;
            }
        }
        if (staged_cnt > 0) {
            push_block(sh, staged, staged_cnt);
            staged_cnt = 0;
        }
    }
    free(raw);
    free(staged);
    record_free(rec);
    return NULL;
}

int audio_start(Audio *a, const char *source, unsigned rate, unsigned channels) {
    if (channels < 1)
        channels = 1;
    if (channels > 2)
        channels = 2;
    a->channels = channels;
    snprintf(a->want_source, sizeof a->want_source, "%s", source);
    a->want_rate = rate;
    a->want_channels = channels;
    if (pthread_create(&a->thread, NULL, capture_thread, a) != 0)
        return 0;
    a->started = 1;
    return 1;
}

size_t audio_consume(Audio *a, const double **left, const double **right) {
    Shared *sh = a->shared;
    size_t frames = 0;
    for (;;) {
        size_t head = atomic_load(&sh->head);
        size_t tail = atomic_load(&sh->tail);
        if (head == tail || frames >= a->cap)
            break;
        size_t slot = tail & sh->mask;
        size_t n = atomic_load(&sh->ring[slot]);
        if (n == (size_t)-1 || n > BLOCK_FRAMES)
            break;
        if (n > 0 && frames + n > a->cap)
            break;
        pthread_mutex_lock(&sh->block_mu[slot]);
        double *blk = sh->blocks + slot * BLOCK_FRAMES * 2;
        for (size_t i = 0; i < n; i++) {
            a->left[frames + i] = blk[i * 2];
            a->right[frames + i] = blk[i * 2 + 1];
        }
        pthread_mutex_unlock(&sh->block_mu[slot]);
        atomic_store(&sh->ring[slot], (size_t)-1);
        atomic_fetch_add(&sh->tail, 1);
        frames += n;
    }
    if (frames > 0) {
        *left = a->left;
        *right = (a->channels > 1) ? a->right : NULL;
    } else {
        *left = NULL;
        *right = NULL;
    }
    return frames;
}

int audio_failed(const Audio *a) {
    return atomic_load(&a->shared->terminate);
}

const char *audio_error(Audio *a) {
    return a->shared->error;
}

void audio_stop(Audio *a) {
    if (!a || !a->started)
        return;
    atomic_store(&a->shared->terminate, 1);
    pthread_join(a->thread, NULL);
    a->started = 0;
}
